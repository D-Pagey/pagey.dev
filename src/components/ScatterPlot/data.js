export default [
    {
        id: '1',
        x: 6,
        y: 6
    },
    {
        id: '2',
        x: 5.5,
        y: 6
    },
    {
        id: '3',
        x: 9,
        y: 2
    },
    {
        id: '4',
        x: 8.5,
        y: 1
    },
    {
        id: '5',
        x: 7,
        y: 3
    },
    {
        id: '6',
        x: 6.5,
        y: 4
    },
    {
        id: '7',
        x: 5.5,
        y: 5
    },
    {
        id: '8',
        x: 9,
        y: 1
    },
    {
        id: '9',
        x: 6,
        y: 4
    }
];
