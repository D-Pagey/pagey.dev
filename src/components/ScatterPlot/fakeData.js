export default [
    {
        id: '1',
        x: 8,
        y: 12
    },
    {
        id: '2',
        x: 2,
        y: 80
    },
    {
        id: '3',
        x: 1,
        y: 95
    },
    {
        id: '4',
        x: 5,
        y: 50
    },
    {
        id: '5',
        x: 6,
        y: 45
    },
    {
        id: '6',
        x: 7,
        y: 15
    },
    {
        id: '7',
        x: 3,
        y: 88
    },
    {
        id: '8',
        x: 4,
        y: 67
    },
    {
        id: '9',
        x: 10,
        y: 58
    }
];
