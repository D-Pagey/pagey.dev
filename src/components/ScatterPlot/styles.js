import styled from 'styled-components';

export const SVG = styled.svg`
    max-width: 500px;
`;
