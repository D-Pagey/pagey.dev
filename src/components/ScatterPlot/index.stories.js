import React from 'react';
import { storiesOf } from '@storybook/react';
import ScatterPlot from '.';

const props = {
    height: 400,
    paddingX: 50,
    paddingY: 35,
    width: 500
};

storiesOf('ScatterPlot', module).add('default', () => <ScatterPlot {...props} />);
