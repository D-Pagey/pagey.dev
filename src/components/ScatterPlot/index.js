import React, { useState } from 'react';
import { scaleLinear } from 'd3-scale';
import * as S from './styles';

import fakeData from './fakeData';

/** TO DO:
 * add all this shit to trello
 * see if can refactor whole component
 * see if can reususe between barchart e.g. axis, constants
 * refactor the scale funcs for BarChart (maybe height func)
 * test coverage
 */

const width = 500;
const height = 400;
const paddingX = 50;
const paddingY = 35;

const ScatterPlot = () => {
    const [dataId, setDataId] = useState();

    const handleHover = (id) => () => setDataId(id);

    const xScale = scaleLinear()
        .domain([0, 10])
        .range([paddingX, width - paddingX]);

    const yScale = scaleLinear()
        .domain([0, 100])
        .range([height - paddingY, 0 + paddingY]);

    return (
        <S.SVG viewBox={`0 0 ${width} ${height}`}>
            <line x1={paddingX} y1={height - paddingY} x2={paddingX} y2={paddingY} stroke="grey" strokeWidth="1" />

            <text x={paddingX - 5} y={paddingY + 5} textAnchor="end">
                100
            </text>

            <text x={paddingX - 50} y={height / 2} transform="rotate(270) translate(-225, -180)">
                yAxisLabel
            </text>

            <text x={paddingX - 5} y={height - paddingY + 5} textAnchor="end">
                0
            </text>

            <text x={width / 2} y={height - 2}>
                xAxisLabel
            </text>

            {/* X axis */}
            <line
                x1={paddingX}
                y1={height - paddingY}
                x2={width - paddingX}
                y2={height - paddingY}
                stroke="grey"
                strokeWidth="1"
            />

            {fakeData.map((item) => (
                <g key={item.id}>
                    <circle
                        cx={xScale(item.x)}
                        cy={yScale(item.y)}
                        r="7"
                        fill="green"
                        onMouseEnter={handleHover(item.id)}
                    />

                    {dataId === item.id && (
                        <>
                            <line
                                x1={paddingX}
                                y1={yScale(item.y)}
                                x2={xScale(item.x)}
                                y2={yScale(item.y)}
                                stroke="red"
                                strokeWidth="1"
                                strokeDasharray="2"
                            />

                            <text x={paddingX - 5} y={yScale(item.y) + 5} textAnchor="end">
                                {item.y}
                            </text>

                            <line
                                x1={xScale(item.x)}
                                y1={yScale(item.y)}
                                x2={xScale(item.x)}
                                y2={height - paddingY}
                                stroke="purple"
                                strokeWidth="1"
                                strokeDasharray="2"
                            />

                            <text x={xScale(item.x) + 4} y={height - 19} textAnchor="end">
                                {item.x}
                            </text>
                        </>
                    )}
                </g>
            ))}
        </S.SVG>
    );
};

export default ScatterPlot;
