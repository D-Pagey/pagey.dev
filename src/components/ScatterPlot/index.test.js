import React from 'react';
import { fireEvent } from '@testing-library/react';
import ScatterPlot from '.';

const props = {
    height: 400,
    paddingX: 50,
    paddingY: 35,
    width: 500
};

describe('ScatterPlot component', () => {
    it('should render', () => {
        const { container } = render(<ScatterPlot {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('should render lines on hover', () => {
        const { getByTestId } = render(<ScatterPlot {...props} />);
        fireEvent.mouseEnter(getByTestId('scatterPlotPoint0'));
        getByTestId('scatterPlotLine0');
    });
});
