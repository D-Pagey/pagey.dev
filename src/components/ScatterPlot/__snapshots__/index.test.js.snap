// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScatterPlot component should render 1`] = `
.c0 {
  max-width: 500px;
}

<div>
  <svg
    class="c0"
    viewBox="0 0 500 400"
  >
    <line
      stroke="grey"
      stroke-width="1"
      x1="50"
      x2="50"
      y1="365"
      y2="35"
    />
    <text
      text-anchor="end"
      x="45"
      y="40"
    >
      10
    </text>
    <text
      transform="rotate(270) translate(-225, -180)"
      x="-25"
      y="200"
    >
      Cups of coffee ‚òïÔ∏è Ô∏è
    </text>
    <text
      text-anchor="end"
      x="45"
      y="370"
    >
      0
    </text>
    <text
      x="190"
      y="398"
    >
      Hours of sleep üò¥
    </text>
    <line
      stroke="grey"
      stroke-width="1"
      x1="50"
      x2="450"
      y1="365"
      y2="365"
    />
    <g>
      <circle
        cx="183.33333333333331"
        cy="167"
        data-testid="scatterPlotPoint0"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="150"
        cy="167"
        data-testid="scatterPlotPoint1"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="383.33333333333337"
        cy="299"
        data-testid="scatterPlotPoint2"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="350"
        cy="332"
        data-testid="scatterPlotPoint3"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="250"
        cy="266"
        data-testid="scatterPlotPoint4"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="216.66666666666669"
        cy="233"
        data-testid="scatterPlotPoint5"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="150"
        cy="200"
        data-testid="scatterPlotPoint6"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="383.33333333333337"
        cy="332"
        data-testid="scatterPlotPoint7"
        fill="green"
        r="9"
      />
    </g>
    <g>
      <circle
        cx="183.33333333333331"
        cy="233"
        data-testid="scatterPlotPoint8"
        fill="green"
        r="9"
      />
    </g>
  </svg>
</div>
`;
